pkg/server/initial_sql.go

```diff
@@ -107,17 +107,28 @@ func (s *topLevelServer) disableReplication(ctx context.Context) (retErr error)
 	if err != nil {
 		return err
 	}
-	// We have to make sure to close the iterator since we might return
-	// from the for loop early (before Next() returns false).
-	defer func() { retErr = errors.CombineErrors(retErr, it.Close()) }()
 
+	// TODO(#125882): We need to cache the zones before we can modfiy them.
+	// This is because the iterator will open a transaction that holds a
+	// a lease to the system database, which will block the ALTER DATABASE
+	// system schema change.
 	var ok bool
+	var zones []string
 	for ok, err = it.Next(ctx); ok; ok, err = it.Next(ctx) {
 		zone := string(*it.Cur()[0].(*tree.DString))
+		zones = append(zones, zone)
+	}
+
+	// We have to make sure to close the iterator since we might return
+	// from the for loop early (before Next() returns false).
+	defer func() { retErr = errors.CombineErrors(retErr, it.Close()) }()
```

Comment:
You need to close the iter before the next step

Reply:
@fqazi i'm not sure i follow: i thought we would want to keep this deferred in case we exit before the iterator continues still -- i don't think calling close is synonymous to a txn completing as well (i could be wrong) since we no longer see that we need to wait for a lease on system to expire

Reply:
Oh its automatically closed on Next, once the rows run out.